### DISCUSSION TITLE
private_key_jwt auth - `aud` property

### DISCUSSION THREAD

**Discussion by leemcmullen** [panva/oauth4webapi#120] [Q&A]
Hi,

I'm trying to get nextauth working with an auth provider who use `private_key_jwt` for authorisation. nextauth uses `oauth4webapi` under the covers.

The issue we're coming across is that when `oauth4webapi` creates the JWT to send to the token endpoint, the `aud` property is being created as an array. The provider we're using will **_strictly only_** accept a string for this property.

The line where this happens is here: https://github.com/panva/oauth4webapi/blob/787efb2ad3ddc9f6b09bb8e229b5c44440b15e22/src/index.ts#L1400

I've found this spec which seems to imply the aud should be a string but then goes on to talk about aud values as a plural, so I don't know for sure: https://datatracker.ietf.org/doc/html/rfc7523#section-2.2

Either way, is it feasible to submit a PR to make aud optionally a string with a single value instead of an array?

Thanks,
Lee

**Comment by panva** [comment]
What is the provider?

**Reply by leemcmullen** [reply]
Hi @panva 

The provider is NHS Login: https://nhsconnect.github.io/nhslogin/oidc-login-flow/

**Reply by leemcmullen** [reply]
Their docs link out to here which seems to imply that aud should be a string: https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

My thinking was we could pass an option in to oauth4webapi to configure it to use a string instead of the default array implementation when creating the JWT.

What do you think?

**Reply by panva** [reply]
Much has been said and written about the `aud` claim in Private Key JWT client assertions.

Bottom line is that the AS should accept both string and array and expect a value that uniquely identify the AS. For historical reasons a single string may be both the issuer identifier or the url of the token endpoint, in an array (which this module does for interoperability reasons) at least one of those needs to be present.

That's as much as we've agreed upon in the FAPI WG and is [summarized in the FAPI 2.0 SP draft](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html#section-5.3.2.1-2.8).

I'd suggest you reach out to NHSConnect first to fix this issue. Their own [interface specification](https://nhsconnect.github.io/nhslogin/interface-spec-doc/) talks about `aud` containing the token endpoint URL which is the correct language since `aud` may be both a string or an array of strings.

![image](https://github.com/panva/oauth4webapi/assets/241506/8cc20516-984e-4149-93fa-7689b0682d2c)

If NHS had undergone [OpenID Connect Certification](https://openid.net/certification/) this would have been identified as a defect in the process.

I could, of course, [add an option](https://github.com/panva/oauth4webapi/compare/main...issue-120) to get out of this interop nightmare once and for all but I'd rather not if it can be avoided.

**Reply by leemcmullen** [reply]
Thanks for your detailed response 👍 

I've been speaking to them in a private slack channel I've been granted access to and they're so far saying that they have no plans to support an array 😕

I'll refer them to this chat but I suspect even if they decide to support it, then it will be a long process and therefore be quite some time before it actually gets implemented and released.

Would it be possible to add an option to override the default and get over this hump? I'd **much much** prefer to continue using the library rather than having to try and work out something different 🙏🙏🙏

If we can get an option to use a string, I'll see about submitting a PR to next-auth to update to the latest version. There's another issue with their lib where they're not even sending the private key into this lib (so private_key_jwt can never work with next-auth at the moment) which I'll look at adding a PR for too.

**Reply by leemcmullen** [reply]
Lol,  just spotted your reply to our chat in the slack channel 👍

**Reply by panva** [reply]
I'm not going to be adding an option just yet on the account of there being an ongoing discussion which I believe NHS should be able and willing to accomodate.

**Reply by leemcmullen** [reply]
@panva I've had an update from NHS in the Slack channel that conversations are now in progress. Hopefully they can make the necessary change 🤞

**Comment by leemcmullen** [comment]
Just following up my comment about nextauth not passing the private key into oauth4webapi, here's the PR for that lib to pass it across: https://github.com/nextauthjs/next-auth/pull/11132

**Reply by balazsorban44** [reply]
I reviewed the PR. Feel free to reach out at https://x.com/balazsorban44 or https://discord.authjs.dev if I don't see your replies.

**Comment by leemcmullen** [comment]
@panva Quick update. The NHS login team have advised (via the Slack channel) that they're going to make the change to the `aud` prop on their side, should be wihtin the next week or so 🎉

Also, my next-auth PR to pass the key into oauth4web has been approved and merged (thanks again @balazsorban44 🙏).
